c/tskit/trees.c

```diff
@@ -4499,10 +4503,10 @@ tsk_diff_iter_next(tsk_diff_iter_t *self, double *ret_left, double *ret_right,
         edges_in.tail = in_tail;
 
         right = sequence_length;
-        if (self->insertion_index < self->num_edges) {
+        if (self->insertion_index < (tsk_id_t) self->num_edges) {
             right = TSK_MIN(right, edges->left[insertion_order[self->insertion_index]]);
         }
-        if (self->removal_index < self->num_edges) {
+        if (self->removal_index < (tsk_id_t) self->num_edges) {
             right = TSK_MIN(right, edges->right[removal_order[self->removal_index]]);
         }
```

Comment:
We need some C tests of this to make sure we're not touching any uninitialised memory. Easiest thing to do is to run through the iterator setting a ``parent`` array as usual, and then check at the end that the parent array is all -1 if TSK_INCLUDE_TERMINAL has been set.

Reply:
Just done this in https://github.com/tskit-dev/tskit/pull/787/commits/b4a3f1d9d26c44b8330bb9bf5276ee7e7c621093 - I've not written many C tests before, so let me know if anything is wrong / unconventional.

Reply:
Nope, these look ideal, thanks (modulo issues @benjeffery points out)